Enhancements: Add Video configuration field for VMs (#36)
kubevirt/enhancements: veps/sig-compute/vm_video_configuration_field.md

# VEP #35: Add a Video configuration field for VMs

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)

# Overview
We want to allow VM-owners to explicitly set the video device type when needed.

## Motivation

> Currently, the default video device type for AMD64 in KubeVirt is **VGA** or **Bochs** in some cases.

Reviewer: The current implementation is already using virtio for ARM [1] and S390x [2].
Only x64 [3] is using bochs for efi, and vga for legacy bios.

[1]
  https://github.com/kubevirt/kubevirt/blob/00418cab40405b8f9aa07238febea4bea942742a/pkg/virt-launcher/virtwrap/converter/arch/arm64.go#L41

[2]
 https://github.com/kubevirt/kubevirt/blob/00418cab40405b8f9aa07238febea4bea942742a/pkg/virt-launcher/virtwrap/converter/arch/s390x.go#L36

[3]
  https://github.com/kubevirt/kubevirt/blob/00418cab40405b8f9aa07238febea4bea942742a/pkg/virt-launcher/virtwrap/converter/arch/amd64.go#L38

Author: adjusted the proposal accordingly

Using `vga` video type can limit the video functionality available to virtual machines.
For example, VGA restricts display resolution settings for Windows VMs and may not support modern workloads effectively.
However, it is necessary to maintain compatibility for older guests that depend on VGA (e.g., RHEL 5/6).

> To address this, we propose introducing a new configuration field to allow users to set the video device type.

Reviewer: To address this, kubevirt is using vga for legacy bios workloads already.

Author: fixed

## Goals

> The VM spec template resource should have the ability to set the video device type.

Reviewer: Why?

## Non Goals

> Change the defaulting logic

Reviewer: Why not change the default logic?

Author: added this option as an alternative to a dedicated video field

## Definition of Users

> VM-owners who require specific video configurations for VMIs in their environments.

Reviewer: Which video adapter configs might required?

Author: adjusted this to
`VM owners who require specific video resolutions that are unavailable when using VGA or Bochs as the video device.
`

## User Stories

> - As a VM owner, I want to be able to set a better video device like `virtio` instead of the standard `vga`.

Reviewer: 1. The user story needs to be platform dependent, because it might be different on ARM, s390x and x64
2. The user story should contain the "so that" or "because" part, see https://en.wikipedia.org/wiki/User_story#Common_templates

Author: adjusted the story to address `amd64` sepcifically, not sure what is needed for arm, and s390x since `virtio` is already enabled there by default.

## Repos

> - [KubeVirt](https://github.com/kubevirt/kubevirt)

Reviewer: Even in a step by step approach, it might be worth to highlight that common-temapltes and common-instancetypes repos might be aftereffect.

## Design

### VM-Level Configuration Field

#### Description
Add a `Video` struct under `spec.template.spec.domain.devices` in the VM template schema.
This struct will include a single field, `type`, to specify the desired video device type.

The current behavior relies on the `autoattachGraphicsDevice` field, where:
1. If `autoattachGraphicsDevice` is not specified or is set to `true`, the current logic in the virt-launcher determines the video device type.
2. If `autoattachGraphicsDevice` is set to `false`, no graphics or video devices are attached.

The proposed change ensures that users can explicitly set the `type` for the video device **only if `autoattachGraphicsDevice` is not explicitly set to `false`**. This constraint will be enforced via the validation webhook and ensures that the new field does not conflict with existing configurations.

The architecture-specific logic will continue to manage the video device type configurations.

> #### Implementation Logic

Reviewer 2: How many new VMs do we expect to use the virtio device?

Author: Do numbers matter? I believe that if even a single user needs this configuration, and we can add it without breaking existing workloads or introducing backward compatibility concerns, we should expand our API to support as many users as possible—especially since the changes are not overly complex or difficult to maintain. What do you think?

> #### Implementation Logic

Reviewer 2: Does this proposal also cover to include the video type in preferences?

Author: Not quite—I prefer a step-by-step approach. First, we should agree on adding this to the API. After that, based on current demand, we can decide whether to open a PR to add it to preferences or instance types.

> #### Implementation Logic

Reviewer 2: Do we have any plan to change the default video device in future?

Author: Fair question—I’d like others to weigh in on this.

In the past, there were concerns about changing the defaulting logic since RHEL 5/6 versions weren’t compatible with virtio (if I recall correctly). However, I’m not sure if we’re required to keep this default forever.

Reviewer 3: It could be a breaking change for existing VMs though. Windows guests could be using hardware details to create unique ids for licensing purposes. Would such ids change then?

Reviewer 3: I am not 100% certain that it would actually change a calculated ID. But it could be possible.

See https://support.microsoft.com/en-us/windows/reactivating-windows-after-a-hardware-change-2c0e962a-f04c-145b-6ead-fb3fc72b6665

Author: I will verify this on a windows machine myself when i get the chance :+1:

1. If `autoattachGraphicsDevice` is set to `false`, the `Video` field must not exist. This will be validated via the webhook.

At the `addGraphicsDevice` method:
2. If `autoattachGraphicsDevice` is set to `false`, no graphics or video devices are attached.
3. If `Video` is explicitly specified in the VM spec, the video type from the spec is used.
4. If `autoattachGraphicsDevice` is not specified or is `true`:
   - Use Bochs for EFI guests on AMD64.
   - Use VGA for BIOS guests on AMD64.

The same hierarchy will apply for other architectures.

> #### Pros and Cons

Reviewer: Pros and Cons related to what?
Aren't there the same alternatives like in https://github.com/kubevirt/enhancements/pull/32 ?
[ In my naive way, I would expect in the design section x a structure like:](https://github.com/kubevirt/enhancements/pull/32#pullrequestreview-2724674671)
x.1 Option/Alternative a

    idea
    advantages
    disadvantages

x.2. Alternative b

    idea
    advantages
    disadvantages

x.3 . Alternative c

    idea
    advantages
    disadvantages

x.4 decision
Approach z is used, because ...

Author: I adjusted the proposal, and added the option for defaulting the logic to set `virtio` video device over `vga/bochs`

Although I am not familiar enough with rhel 5/6 machines to determine weather the the backward compatibility concerns are valid or not.

Reviewer: > I adjusted the proposal, and added the option for defaulting the logic to set virtio video device over vga/bochs

Thanks, it improved the readability to me significantly!

> Although I am not familiar enough with rhel 5/6 machines to determine weather the the backward compatibility concerns are valid or not.

These old legacy guests are using BIOS, and I would vote for not touching VGA for BIOS default config.

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)

# Overview
We want to allow VM-owners to explicitly set the video device type when needed.

## Motivation
In KubeVirt, the default video device type for AMD64 architectures is VGA for BIOS-based VMs and Bochs for EFI-based VMs.
For Arm and s390x architectures, we have transitioned to using the virtio video device.
The use of VGA/Bochs can impose limitations on video functionality, such as restricted display resolution settings, particularly for Windows virtual machines.
However, maintaining compatibility with older guest operating systems (e.g., RHEL 5/6) that rely on VGA/Bochs remains important.
To balance modernization with backward compatibility, KubeVirt continues to use VGA for legacy BIOS-based workloads.

## Goals
Ensuring VM-owners won't have a limited resolution experience with their VMs

## Definition of Users
VM owners who require specific video resolutions that are unavailable when using VGA or Bochs as the video device.

## User Stories

> As an AMD64 VM owner, I want my VM to provide high-resolution display settings when accessed via RDP, VNC,

Reviewer: RDP has its own virtual display adapter.
```suggestion
As an AMD64 VM owner, I want my VM to provide high-resolution display settings when accessed via VNC
```

Author: done

or other remote access methods, so that I have a better user experience when working with graphical applications.

## Repos
- [KubeVirt](https://github.com/kubevirt/kubevirt)
- [common-templates](https://github.com/kubevirt/common-templates)
- [common-instancetypes](https://github.com/kubevirt/common-instancetypes)

## Design

> ### Option 1: Default to Virtio Instead of VGA/Bochs

Reviewer: Technically, this is correct.
for prgamtic reasons, I propose to focus this option only on uefi:
```suggestion
### Option 1: Default to Virtio Instead of Bochs for UEFI VMs
```

Author: fixed

#### Description
Instead of relying on VGA/Bochs as the default video device for AMD64 architectures, 
KubeVirt would transition to virtio as the new default across all architectures.
This change would provide better video performance, higher resolutions,
and improved compatibility with modern operating systems,especially Windows VMs.

**Pros:**
- **Reduced Configuration:** Users won't need to manually switch to virtio for better video performance.

**Cons:**

> - **Backward Compatibility:** we might need to adjust a fall-back logic to maintain full backward-computability

Reviewer: This is a very valid point, and we should think twice about it.
I noticed some difference in bochs to virtio-vga, e.g. 8mb vs 16mb mem, the way the default resolution form OVMF is applied. But I did not notice anything what worries me, but it does not mean that there is no issue I did not recognize.

The architecture-specific logic will continue to manage the video device type configurations.

#### Implementation Logic
1. If `autoattachGraphicsDevice` is set to `false`, the `Video` field must not exist. This will be validated via the webhook.

At the `addGraphicsDevice` method:
2. If `autoattachGraphicsDevice` is set to `false`, no graphics or video devices are attached.
3. If `Video` is explicitly specified in the VM spec, the video type from the spec is used.
4. If `autoattachGraphicsDevice` is not specified or is `true`:
   - Use Bochs for EFI guests on AMD64.
   - Use VGA for BIOS guests on AMD64.

The same hierarchy will apply for other architectures.

#### Pros and Cons

**Pros:**
- **Granular Control:** Enables precise configuration for each VM.

**Cons:**
- **New API:** Introducing additional fields that need maintenance and testing.

### API Examples

#### Video Device Configured Explicitly 
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
   name: vm
spec:
   template:
      spec:
         domain:
            devices:
               video:
                  type: virtio
```

> 

Reviewer: Option 3: Use an inoffical API like an annotation to incluence the behavior of kubevirt

This proposal suggests to add the kubevirt.io/virtio-vga annotation to VirtualMachine objects, which will trigger KubeVirt to make the required configuration changes to enable virtio graphics adapter for the VM. By using an annotation the impact to the core API is kept to a minimum and there is room for future changes to this feature independently of the core API.
Other way round, an annotation like ` kubevirt.io/bochs` could be used to ensure backwards compatibility in case the change to virtio intruduced regressions. In a view releases, the anotation could be removed, without conflicting with the API versioning rules.

Author: Added this alternative

# VEP #35: Add a Video configuration field for VMs

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)

# Overview
We want to allow VM-owners to explicitly set the video device type when needed.

## Motivation
In KubeVirt, the default video device type for AMD64 architectures is VGA for BIOS-based VMs and Bochs for EFI-based VMs.
For Arm and s390x architectures, we have transitioned to using the virtio video device.
The use of VGA/Bochs can impose limitations on video functionality, such as restricted display resolution settings, particularly for Windows virtual machines.
However, maintaining compatibility with older guest operating systems (e.g., RHEL 5/6) that rely on VGA/Bochs remains important.
To balance modernization with backward compatibility, KubeVirt continues to use VGA for legacy BIOS-based workloads.

## Goals

> Ensuring VM-owners won't have a limited resolution experience with their VMs

Reviewer 3: Is it all about resolution? The next issue could be that the performance of emulated video devices is too slow for VDI purposes.

Author: could you elaborate on VDI?

Reviewer 3: If the aim of this proposal is to encourage the use of KubeVirt as a VDI solution then the next issue a user might encounter is that the performance of an emulated graphics device could be too slow for use with certain applications that rely on hardware acceleration.

## Definition of Users

> VM owners who require specific video resolutions that are unavailable when using VGA or Bochs as the video device.

Reviewer 3: Why are they unavailable? Missing device drivers?

Author: It's because the resolution is being set to a default value 1280X800, and we don't offer api for adjusting the resolution

## User Stories
As an AMD64 VM owner, I want my VM to provide high-resolution display settings when accessed via VNC

> or other remote access methods, so that I have a better user experience when working with graphical applications.

Reviewer 3: See comment above, will it really improve the user experience for VDI purposes?

Author: Definitely

Reviewer 2: KubeVirt is no VDI solution.

VDI includes graphics, but also multi monitor, tablets, usb passthrough, audio, video encoding. VNC does not provide all of this.

Thus let me echo Felix: How does setting the device to virtio help? Please provide some details.

Reviewer 4: > KubeVirt is no VDI solution.

Sure, let's get rid of this argument then.

> How does setting the device to virtio help? Please provide some details.

IIUC the main advantages of virtio are generally better performance (no need to emulate a "real" graphics device) and better support for higher resolutions and configurations.

Is there anything else you expect as a benefit?

Reviewer 3: We don't have customers, we have users. :)

I agree, let's remove the wording around VDI, just state that it allows to specify what type of video device is going to be used and why this is beneficial, i.e. better guest support or better support for higher resolutions and configurations.

Reviewer: I understood that this is about in-guest remote display mirroring, which would be reflected by a user story like
```suggestion
- As an AMD64 VM owner, I want my VM to provide high-resolution display, so that I can use high resolution when I access the VM display via an in-guest graphical remote access methods like an a in guest VNC server```

Author: Thanks, I adjusted the design.

## Repos
- [KubeVirt](https://github.com/kubevirt/kubevirt)
- [common-templates](https://github.com/kubevirt/common-templates)
- [common-instancetypes](https://github.com/kubevirt/common-instancetypes)

## Design

### Option 1: Default to Virtio Instead of Bochs for UEFI VMs
#### Description
Instead of relying on `Bochs` as the default video device for `AMD64` architectures, 
KubeVirt would transition to `virtio` as the new default across all architectures.
This change would provide better video performance, higher resolutions,
and improved compatibility with modern operating systems,especially Windows VMs.

**Pros:**
- **Reduced Configuration:** Users won't need to manually switch to virtio for better video performance.

**Cons:**

> - **Backward Compatibility:** we might need to adjust a fall-back logic to maintain full backward-computability

Reviewer 3: See my comment about unique ids derived from the hardware.

Author: Didn't find a source that say replacing video device will result uuid calculation to change. but it doesn't mean it's not true.. also if we go with api configuration rather then switching the default, on cases where the uuid calc does break, it will only affect the specific users who want to have virtio video over the default one, in such case we can throw a warning or let them know the uuid might break

Reviewer 3: You could test it with a Windows VM to be sure.

Author: what about the firmware uuid? we are the ones that generates and persist it, it wont be change due to the video device switch

Reviewer 3: Yeah, let's move forward ruling out this alternative.

@dasionov We control the firmware UUID, but we don't control what Windows takes into account for computing a "hardware ID". That's why I was suggesting to simply test it.

Author: Thank you for clarifying, now I understand it better. :+1:

Author: adjusted the design to mark this alternative was ruled out due to compatibility issues

### Option 2 VM-Level Configuration Field

#### Description
Add a `Video` struct under `spec.template.spec.domain.devices` in the VM template schema.
This struct will include a single field, `type`, to specify the desired video device type.

The current behavior relies on the `autoattachGraphicsDevice` field, where:
1. If `autoattachGraphicsDevice` is not specified or is set to `true`, the current logic in the virt-launcher determines the video device type.
2. If `autoattachGraphicsDevice` is set to `false`, no graphics or video devices are attached.

> The proposed change ensures that users can explicitly set the `type` for the video device **only if `autoattachGraphicsDevice` is not explicitly set to `false`**. This constraint will be enforced via the validation webhook and ensures that the new field does not conflict with existing configurations.

Reviewer 3: Will it allow to set `spec.template.spec.domain.devices` if `autoattachGraphicsDevice` is set to `true`?

Author: yes, IIRC the code in the converter won't add the graphic device only when `autoattachGraphicsDevice` is != nil and `autoattachGraphicsDevice` == false
 
 so when it's true it should be ok.

## User Stories
As an AMD64 VM owner, I want my VM to provide high-resolution display settings when accessed via VNC
or other remote access methods, so that I have a better user experience when working with graphical applications.

## Repos
- [KubeVirt](https://github.com/kubevirt/kubevirt)
- [common-templates](https://github.com/kubevirt/common-templates)
- [common-instancetypes](https://github.com/kubevirt/common-instancetypes)

## Design

### Option 1: Default to Virtio Instead of Bochs for UEFI VMs
#### Description
Instead of relying on `Bochs` as the default video device for `AMD64` architectures, 
KubeVirt would transition to `virtio` as the new default across all architectures.
This change would provide better video performance, higher resolutions,
and improved compatibility with modern operating systems,especially Windows VMs.

**Pros:**
- **Reduced Configuration:** Users won't need to manually switch to virtio for better video performance.

**Cons:**
- **Backward Compatibility:** we might need to adjust a fall-back logic to maintain full backward-computability

### Option 2 VM-Level Video Configuration Field

#### Description

> Introduce a Video struct under spec.template.spec.domain.devices in the VM template schema.

Reviewer: Is there the assumption that there can be only a single `Video` ? Should the API prepare for having multiple ones?

Author: at this moment i would prefer to have a single configuration just for primary video, if the use-cases in the feature rises to support more then the primary configuration it can be added later wdyt?

Reviewer 3: TBH I'm not sure whether such an API will benefit us in the long term. Maybe an annotation to force virtio GPUs is enough for now? What do you think @vladikr?

Reviewer: I would vote for the other way round, using virtio by default, especially virtio-vga on x64. If this results in trouble, an annotation to fall back on bochs would give us time to understand why the backward compatibility issue happened and if an explicit API is required at all.

Reviewer 3: I don't think we can change the default for now, see the other comment threads.

Author: Agreed.

The current behavior relies on the `autoattachGraphicsDevice` field, where:
1. If `autoattachGraphicsDevice` is not specified or is set to `true`, the current logic in the virt-launcher determines the video device type.
2. If `autoattachGraphicsDevice` is set to `false`, no graphics or video devices are attached.

The proposed change ensures that users can explicitly set the `type` for the video device **only if `autoattachGraphicsDevice` is not explicitly set to `false`**. This constraint will be enforced via the validation webhook and ensures that the new field does not conflict with existing configurations.

The architecture-specific logic will continue to manage the video device type configurations.

#### Implementation Logic
1. If `autoattachGraphicsDevice` is set to `false`, the `Video` field must not exist. This will be validated via the webhook.

At the `addGraphicsDevice` method:
2. If `autoattachGraphicsDevice` is set to `false`, no graphics or video devices are attached.
3. If `Video.Type` is explicitly specified in the VM spec, the video type from the spec is used.

> 4. If `Video.Resolution` is explicitly specified the video resolution from the spec is used.

Reviewer: Just a comment: There might be dependencies in both directions between `Video.Type` and `Video.Resolution`, not sure if this can be avoided.

Author: yes, you are correct, if a user will provide a video type which supports only resolution for 1920X1080, and configure resolution for 4K it should result an error

Reviewer 3: How do we know that?

Author: I believe we should have minimal validation regarding this, let the errors be reported from libvirt in case of configuration miss match, like unsupported resolution per specific video. 

so we start with validating:
`devices.video` != nil -> `autoattachGraphicsDevice != false`
and since resolution only supported for `vga`, `qxl`, `bochs`, `gop`, and `virtio`
we can validate resolution != nil if video != nil and video.type in  (`vga`, `qxl`, `bochs`, `gop`, `virtio`)

The same hierarchy will apply for other architectures.

#### Pros and Cons

**Pros:**

> - **Granular Control:** Enables precise configuration for each VM.

Author: The user story highlights the limited resolution issue that Windows users are facing.
By adding a resolution setting to the API, we will be able to address the problem described in the user story.

> - **Granular Control:** Enables precise configuration for each VM.

Reviewer: The problem would be already addressed if the default would change to virtio.

> - **Granular Control:** Enables precise configuration for each VM.

Author: maybe, but we can't switch the default... I guess even for EFI VMs since virtio require driver to be opt-in.. we cant afford to break efi vms without prepared drivers.
@xpivarc @jean-edouard i initially thought it was possible to request virtio video device, which is processed by libvirt as `virtio-gpu` and if the drivers are not exist we will fallback to `virtio-vga`
but i am not sure this is possible anymore

wdyt?

> - **Granular Control:** Enables precise configuration for each VM.

Reviewer 5: IIRC, @dasionov you tried booting a diskless EFI VM with virtio graphics and didn't see the firmware screen(s). That on its own is problematic enough to avoid switching to virtio by default IMO

> - **Granular Control:** Enables precise configuration for each VM.

Reviewer: sure, there might be questions to investigate, e.g. if virtio-vga was used in this case, or if something like in https://www.kraxel.org/blog/2019/02/ramfb-display-in-qemu/ should be used.

> - **Granular Control:** Enables precise configuration for each VM.

Reviewer 3: > IIRC, @dasionov you tried booting a diskless EFI VM with virtio graphics and didn't see the firmware screen(s). That on its own is problematic enough to avoid switching to virtio by default IMO

That means option one is the preferred choice?

> - **Granular Control:** Enables precise configuration for each VM.

Author: The preferred choice is adding API for VM Video configuration, I adjusted the design to clarify this better.

In other words, the video device that is chosen for VMs is implicit and unclear.
Different video devices have pros and cons and fit to different use-cases.
For example, a guest with `virtio` drivers installed could enjoy from a `virtio` device which is generally faster and optimized for virtualization.
Legacy guests, or guests with no `virtio` drivers, can use VGA for compatibility, and so on.

With this proposal, the VM creator could adapt the video deceive to their needs.

## Goals
- Allowing VM-owners to choose `virtio` or any other desired video device to override the existing default one.

## Definition of Users
- VM owners who need to select a specific video device for their virtual machines to match their workload requirements.

## User Stories
- As an VM owner, I want my VM to provide high-resolution display, so that I can use high resolution
  when I access the VM display via an in-guest graphical remote display mirroring access method like an in guest VNC server

- As a Windows 11 on ARM user, I like to install my VM using the ramfb display adapter,
  and replace it by the `virtio` adapter after I installed the `virtio` driver

> - as a VM creator with legacy guest or guest with no `virtio` drivers, I want to rely on VGA.

Reviewer: ```suggestion
- As a VM creator with legacy guest or guest with no `virtio` drivers, I want to rely on VGA for my x64 VMs.
```
Please note if you request virtio from libvirt on x64, you will get virtio-vga which provides vga compatibility.
An non-x64 platforms there is no VGA at all.

Author: even when drivers are already installed?

Reviewer: > even when drivers are already installed?

I do not understand the question. kubevirt creates xml which is send to libvirt which is generting the qemu command line from it. So if kubevirt generates xml to use virtio, libvirt generates a virtio-vga command line for qemu. This all happens before the vm is started and the driver is relevant.

Author: sorry, you are right, but virtio-vga would still work for the goal of this vep will it not?

> - as a VM creator running a guest with `virtio` drivers, I want to enjoy from an optimized virtualization-aware video device which is generally faster and lighter on resources.

Reviewer: ```suggestion
- As a VM creator running a guest with `virtio` drivers, I want to enjoy from an optimized virtualization-aware video device which is generally faster and lighter on resources.
```

Reviewer: I am not convicned if this is significant here, but it might not hurt.

Author: I agree

# VEP #35: Add a Video configuration field for VMs

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)

# Overview

> We want to allow VM-owners to explicitly set the video device type when needed.

Reviewer 6: Don't we want 2 things? Resolution and video type?

Author: True, fixed

## Motivation
In KubeVirt, the default video device type for AMD64 architectures is VGA for BIOS-based VMs and Bochs for EFI-based VMs.
For Arm and s390x architectures, we have transitioned to using the virtio video device.

> The use of VGA/Bochs can impose limitations on video functionality, such as restricted display resolution settings, particularly for Windows virtual machines.

Reviewer 6: I believe this sentence is not true.

Author: removed this sentence

However, maintaining compatibility with older guest operating systems (e.g., RHEL 5/6) that rely on VGA/Bochs remains important.

> To balance modernization with backward compatibility, KubeVirt continues to use VGA for legacy BIOS-based workloads.

Reviewer 6: The latter part of the sentence is probably still tide to the "defaulting" VEP. This should be removed as it is not relevant imo

Author: yes, removed it

## Goals

> Ensuring VM-owners won't have a limited resolution experience with their VMs

Reviewer 6: Do you want to add: User have option to choose virtio?

Author: Added, Thanks

## Definition of Users
- VM owners whose workloads require specific display resolutions but are unable to modify them at runtime. 

> In these cases, the guest operating system defaults to a limited resolution during initialization due to the constraints of the default VGA or Bochs video devices.

Reviewer 6: What are the constraints?

Author: removed this part.

> - Users running graphical applications that benefit from higher resolution and better performance, 

Reviewer 3: To me that's a problematic use case. virtio GPUs are emulated on the CPU and not very well suited for high performance applications.

Author: removed the problematic use case as our priority for now reside in addressing the resolution concern

particularly those relying on hardware-accelerated graphics. 
The current Bochs configuration lacks the necessary support for such use cases, especially in modern desktop environments or remote access scenarios like VNC.

> ## User Stories

Reviewer 6: Another one could be:
Application requiring specific resolution in order to work. Might not be for access/VDI, it could be for testing some application and others...

Author: added this use case

As an AMD64 VM owner, I want my VM to provide high-resolution display settings when accessed via VNC
or other remote access methods, so that I have a better user experience when working with graphical applications.

## Repos
- [KubeVirt](https://github.com/kubevirt/kubevirt)
- [common-templates](https://github.com/kubevirt/common-templates)
- [common-instancetypes](https://github.com/kubevirt/common-instancetypes)

## Design

> ### Option 1: Default to Virtio Instead of Bochs for UEFI VMs

Reviewer 6: We need to have a design here, not options and alternatives (and why your or after review, those options are alternatives)

Reviewer: > not options and alternatives
Why not?
Can't we list the options, and let the approver provide a subsection in the end which says "option x is choosen because of ...." , because the approver decides, not the author of the PR.

Reviewer 3: I'm missing the preferred option in this VEP right now. This doc should contain that piece of information. We can list alternatives, but there has to be a chosen option. Documenting it only as a comment on the PR is not enough in my opinion.

Author: refactored the design to clarify the preferred option, which is to extand the api to allow specific configuration for video device and resolution

added a section under it for the alternative options

#### Description
Instead of relying on `Bochs` as the default video device for `AMD64` architectures, 
KubeVirt would transition to `virtio` as the new default across all architectures.
This change would provide better video performance, higher resolutions,
and improved compatibility with modern operating systems,especially Windows VMs.

**Pros:**
- **Reduced Configuration:** Users won't need to manually switch to virtio for better video performance.

**Cons:**

> - **Backward Compatibility:** we might need to adjust a fall-back logic to maintain full backward-computability

Reviewer 6: `we might need` -> `we need`

Author: done

### Option 2 VM-Level Video Configuration Field

#### Description
Introduce a Video struct under spec.template.spec.domain.devices in the VM template schema.
The struct includes:
* type: optionally to specifies the video device type (e.g., virtio, vga, bochs).

> * resolution: An optional sub-struct with x and y fields to define the display resolution (e.g., x: 1920, y: 1080).

Reviewer 6: What would be valid options?

Author: IMO either asking for virtio video model, or asking for a specific resolution with the default video device

The current behavior relies on the `autoattachGraphicsDevice` field, where:
1. If `autoattachGraphicsDevice` is not specified or is set to `true`, the current logic in the virt-launcher determines the video device type.
2. If `autoattachGraphicsDevice` is set to `false`, no graphics or video devices are attached.

The proposed change ensures that users can explicitly set the `type` for the video device **only if `autoattachGraphicsDevice` is not explicitly set to `false`**. This constraint will be enforced via the validation webhook and ensures that the new field does not conflict with existing configurations.

> The architecture-specific logic will continue to manage the video device type configurations.

Reviewer 6: The validation will also needs to take care of architectures

Author: yes, fixed

**Pros:**
- **No API Changes:** Avoids modifying the VM schema.
- **Flexible:** Can be easily removed or changed in future releases without API versioning concerns.

**Cons:**
- **Lack of Visibility and Transparency:** Using an annotation for domain-level configuration makes it harder for users 
to discover and understand its impact. 
Unlike structured API fields, annotations are not well-documented in API references or validation mechanisms.

- **Potential for Unintended Overrides:** Since annotations are loosely enforced,
- they could create inconsistencies if users unknowingly override expected defaults.

### API Examples

#### Video Device Configured Explicitly with an API
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
   name: vm
spec:
   template:
      spec:
         domain:
            devices:
               video:
                  type: virtio
                  resolution:
                     x: 1920
                     y: 1080
```

#### Video Device Configured With Annotation
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
   name: vm
annotations:
  kubevirt.io/video-device: "virtio"
```

## Scalability

> Overhead of the virt-launcher/qemu may be impacted.

Reviewer 6: We need to explicitly state that this will be measured and the overhead will be adjusted

Author: done

# VEP #35: Add a Video configuration field for VMs

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)

# Overview
- We want to allow VM-owners to explicitly set the video device type when needed.

> - We want to allow VM-owners to explicitly set the desired video resolution to overcome the default resolution defined by libvirt/qemu

Reviewer 4: leftover?

Author: kinda, but i chose to leave it at that because it's an issue being solved by switching the video device type, and not by introducing new api for resolution anymore

Author: will drop it

Author: done

## Motivation

> In KubeVirt, the default video device type for AMD64 architectures is VGA for BIOS-based VMs and Bochs for EFI-based VMs.
> For Arm and s390x architectures, we have transitioned to using the `virtio` video device.

Reviewer 4: I don't see the motivation here.
Can you please write something like:
```suggestion
In KubeVirt, the default video device type for AMD64 architectures is VGA for BIOS-based VMs and Bochs for EFI-based VMs.
For Arm and s390x architectures, we have transitioned to using the `virtio` video device.

In other words, the video device that is chosen for VMs is implicit and unclear.
Different video devices have pros and cons and fit to different use-cases.
For example, a guest with virtio drivers installed could enjoy from a virtio device which is generally faster and optimized for virtualization. Legacy guests, or guests with no virtio drivers, can use VGA for compatibility, and so on.

With this proposal, the VM creator could adapt the video deceive to their needs.
```

## Goals

> - Allowing VM-owners to have control over the resolution of the video device with their VMs

Reviewer 4: leftover?

Author: removed it

- Allowing VM-owners to choose `virtio` or any other desired video device to override the existing default one.

## Definition of Users

> - VM owners whose workloads require specific display resolutions but are unable to modify them at runtime. 

Reviewer 4: leftover.
You can rephrase the motivation part into here.

Author: done.

  In these cases, the guest operating system defaults to a limited resolution during boot.

> ## User Stories

Reviewer 4: I'd add:
* as a VM creator with legacy guest or guest with no virtio drivers, I want to rely on VGA.
* as a VM creator running a guest with virtio drivers, I want to enjoy from an optimized virtualization-aware video device which is generally faster and lighter on resources.

These are the most simple user stories here AFAICT

Author: done, thanks.

> - As VM owner I want to run applications requiring a specific resolution in order to work

Reviewer 4: leftover

Author: removed

- As an VM owner, I want my VM to provide high-resolution display, so that I can use high resolution
  when I access the VM display via an in-guest graphical remote display mirroring access method like an in guest VNC server

- As a Windows 11 on ARM user, I like to install my VM using the ramfb display adapter,
  and replace it by the virtio adapter after I installed the virtio driver

## Repos
- [KubeVirt](https://github.com/kubevirt/kubevirt)
- [common-templates](https://github.com/kubevirt/common-templates)
- [common-instancetypes](https://github.com/kubevirt/common-instancetypes)

## Design

> ### Preferred Option: Add `Video` Configuration Field to VM API

Reviewer 4: Please mention that this approach was chosen to implement.

Author: done

Unlike structured API fields, annotations are not well-documented in API references or validation mechanisms.

- **Potential for Unintended Overrides:** Since annotations are loosely enforced,
- they could create inconsistencies if users unknowingly override expected defaults.

### API Examples

#### Video Device Configured Explicitly with an API
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
   name: vm
spec:
   template:
      spec:
         domain:
            devices:
               video:
                  type: virtio
```

#### Video Device Configured With Annotation
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
   name: vm
annotations:
  kubevirt.io/video-device: "virtio"
```

## Scalability
Overhead of the virt-launcher/qemu will be measured and adjusted accordingly.

## Update/Rollback Compatibility
Currently, the virt-launcher arch-converter manages the default video type.
We only provide a field for the user to specify explicitly if they want something else like virtio, so backward compatibility isn't affected.

## Functional Testing Approach
* Create a VM with video.Type set to `virtio` and expect launch successfully.

> # Implementation Phases

Reviewer 4: This is section is for implementation phases that had already happened. IOW you'll add major PRs to here after they'll get merged.

So you can remove this section for now, or move the important stuff to the alpha requirements section.

Author: removed this block.

# VEP #35: Add a Video configuration field for VMs

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)

# Overview
- We want to allow VM-owners to explicitly set the video device type when needed.
- We want to allow VM-owners to explicitly set the desired video resolution to overcome the default resolution defined by libvirt/qemu

## Motivation
In KubeVirt, the default video device type for AMD64 architectures is VGA for BIOS-based VMs and Bochs for EFI-based VMs.
For Arm and s390x architectures, we have transitioned to using the `virtio` video device.

## Goals
- Allowing VM-owners to have control over the resolution of the video device with their VMs
- Allowing VM-owners to choose `virtio` or any other desired video device to override the existing default one.

## Definition of Users

> - VM owners whose workloads require specific display resolutions but are unable to modify them at runtime. 
> In these cases, the guest operating system defaults to a limited resolution during boot.

Reviewer 3: ```suggestion
- VM owners whose workloads require specific display resolutions but are unable to modify them at runtime. 
  In these cases, the guest operating system defaults to a limited resolution during boot.
```

Author: done.

## User Stories
- Application requiring specific resolution in order to work

- As an AMD64 VM owner, I want my VM to provide high-resolution display, so that I can use high resolution 

> when I access the VM display via an in-guest graphical remote display mirroring access methods like an in guest VNC server

Reviewer 3: ```suggestion
when I access the VM display via an in-guest graphical remote display mirroring access method like an in guest VNC server
```

Author: done, fixed.

#### Description
Introduce a structured Video field under spec.template.spec.domain.devices with the following optional attributes:
```yaml
video:
  type: virtio         # Optional: virtio, bochs, vga, etc.
  resolution:          # Optional
    x: 1920
    y: 1080
```
This approach extends the API to provide declarative control over video settings.

#### Integration with Existing Behavior:
* If `autoattachGraphicsDevice` is set to `false`, no video devices will be attached, and the video field must not be present.
* If `video.type` is provided, it will override the default video device model.
* If `video.resolution` is provided, it will be used.
* Webhook will validate that `video` is only set if `autoattachGraphicsDevice` is true or unset, and is valid with the relevant arch.
* Architecture-specific logic will continue to provide sensible defaults unless overridden.

The same hierarchy will apply for other architectures.

#### Pros and Cons

**Pros:**
- **Granular Control:** Enables precise configuration for each VM.

**Cons:**
- **New API:** Introducing additional fields that need maintenance and testing.

## Alternatives Considered:

> ### Option 1: Default to Virtio Instead of Bochs for UEFI VMs - [Ruled out due to compatibility issues]

Reviewer: We did not check if there are compatibility issues, the issues are just a theoretical possibility which we do not know if they could happen.
```suggestion
### Option 1: Default to Virtio Instead of Bochs for UEFI VMs - [Ruled out due to possible compatibility issues]
```

Author: Thanks, fixed.

#### Description
Instead of relying on `Bochs` as the default video device for `AMD64` architectures, 
KubeVirt would transition to `virtio` as the new default across all architectures.
This change would provide better video performance, higher resolutions,

> and improved compatibility with modern operating systems,especially Windows VMs.

Reviewer 2: This is not not always true.
The pre-requisite is the guest to have the relevant drivers installed.

Author: added pre-requisite as a con to the alternative and adjusted the design.

**Pros:**

> - **Reduced Configuration:** Users won't need to manually switch to virtio for better video performance.

Reviewer 2: Maybe virtio is leading to better in VM performance, but from an e2e perspective this is not necessarily given: VNC is not a great protocol for high resolutions, by default VNC traffic goes thorugh the API server, there high traffic is not desired.

Author: True, and also this alternative was ruled out due to compatibility reasons

**Cons:**
- **Backward Compatibility:** we need to adjust a fall-back logic to maintain full backward-computability

> - **pre-requisite driver** `virtio` driver must be installed for `virtio` video to work properly

Author: Thanks! adjusted the design to your suggestion

# VEP #35: Add a Video configuration field for VMs

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)

# Overview
- We want to allow VM-owners to explicitly set the video device type when needed.
- We want to allow VM-owners to explicitly set the desired video resolution to overcome the default resolution defined by libvirt/qemu

## Motivation
In KubeVirt, the default video device type for AMD64 architectures is VGA for BIOS-based VMs and Bochs for EFI-based VMs.
For Arm and s390x architectures, we have transitioned to using the `virtio` video device.

## Goals
- Allowing VM-owners to have control over the resolution of the video device with their VMs
- Allowing VM-owners to choose `virtio` or any other desired video device to override the existing default one.

## Definition of Users
- VM owners whose workloads require specific display resolutions but are unable to modify them at runtime. 
In these cases, the guest operating system defaults to a limited resolution during boot.

## User Stories

> - Application requiring specific resolution in order to work

Reviewer 3: ```suggestion
- As VM owner I want to run applications requiring a specific resolution in order to work
```

Author: fixed, thanks.

> - As an AMD64 VM owner, I want my VM to provide high-resolution display, so that I can use high resolution 

Reviewer 3: ```suggestion
- As an VM owner, I want my VM to provide high-resolution display, so that I can use high resolution 
```

Author: fixed

when I access the VM display via an in-guest graphical remote display mirroring access methods like an in guest VNC server

- As a Windows 11 on ARM user, I like to install my VM using the ramfb display adapter,
and replace it by the virtio adapter after I installed the virtio driver

## Repos
- [KubeVirt](https://github.com/kubevirt/kubevirt)
- [common-templates](https://github.com/kubevirt/common-templates)
- [common-instancetypes](https://github.com/kubevirt/common-instancetypes)

## Design

### Preferred Option: Add `Video` Configuration Field to VM API

#### Description
Introduce a structured Video field under spec.template.spec.domain.devices with the following optional attributes:
```yaml
video:
  type: virtio         # Optional: virtio, bochs, vga, etc.
  resolution:          # Optional
    x: 1920

>     y: 1080

Reviewer 3: I'd drop this, video type should be enough for a start. We don't want to bloat the API too much.

Reviewer 2: +1

I also think that this can be controlled from within the guest. Maybe we want to set a modern resolution by default.

Author: done, adjusted the design and dropped this addition

## Update/Rollback Compatibility
Currently, the virt-launcher arch-converter manages the default video type.
We only provide a field for the user to specify explicitly if they want something else like virtio, so backward compatibility isn't affected.

## Functional Testing Approach
* Create a VM with video.Type set to `virtio` and expect launch successfully.
* Create a VM with video.Resolution set to `1920x1080` and expect this resolution in the guest.

# Implementation Phases
1. **Phase 1: Introduce New Field to Allow Users to Explicitly Set Their Desired Video Type**
    - Add logic to the validation webhook to prevent the creation of VMs where `Video` is specified and `autoattachGraphicsDevice` is explicitly set to `false`.
    - Add logic in the virt launcher to check the field's existence at `addGraphicsDevice`, before setting video device (virtio/vga).
    - Write unit and functional tests for the new behavior.

2. **Phase 2 Documentation**
    - Update KubeVirt documentation to include examples of configuring the video device type.

## Feature Stage

We are targeting this feature to be introduced in **Alpha**.

- During Alpha:
   - The `video` field will be **optional**.

>    - The feature will be **enabled by default**.

Reviewer 3: How so? If no video device was specified what will happen?

Reviewer 4: Alpha features shouldn't be on by default, by rely on a feature gate.
Please remove this point.

Author: adjusted to ` - The feature will be **disabled by default**.`
Thanks for the update !
